<a href="https://colab.research.google.com/github/Gocrazy53/InternshipBootcamp/blob/main/05_multiple_regression_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
house = pd.read_csv('https://github.com/YBI-Foundation/Dataset/raw/main/Boston.csv')
house.head()

CRIM    ZN  INDUS  CHAS     NX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [3]:
house.describe()

CRIM          ZN       INDUS        CHAS          NX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

In [4]:
house.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NX       506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


In [5]:
house.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT', 'MEDV'],
      dtype='object')

In [6]:
y = house['MEDV']
x = house.drop(['MEDV'],axis=1)

In [7]:
# train test split
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,train_size = 0.7,random_state = 2529)

In [8]:
x_train,x_test,y_train,y_test

(        CRIM    ZN  INDUS  CHAS     NX     RM   AGE      DIS  RAD    TAX  \
 310  2.63548   0.0   9.90     0  0.544  4.973  37.8   2.5194    4  304.0   
 202  0.02177  82.5   2.03     0  0.415  7.610  15.7   6.2700    2  348.0   
 76   0.10153   0.0  12.83     0  0.437  6.279  74.5   4.0522    5  398.0   
 152  1.12658   0.0  19.58     1  0.871  5.012  88.0   1.6102    5  403.0   
 186  0.05602   0.0   2.46     0  0.488  7.831  53.6   3.1992    3  193.0   
 ..       ...   ...    ...   ...    ...    ...   ...      ...  ...    ...   
 228  0.29819   0.0   6.20     0  0.504  7.686  17.0   3.3751    8  307.0   
 399  9.91655   0.0  18.10     0  0.693  5.852  77.8   1.5004   24  666.0   
 316  0.31827   0.0   9.90     0  0.544  5.914  83.2   3.9986    4  304.0   
 50   0.08873  21.0   5.64     0  0.439  5.963  45.7   6.8147    4  243.0   
 352  0.07244  60.0   1.69     0  0.411  5.884  18.5  10.7103    4  411.0   
 
      PTRATIO       B  LSTAT  
 310     18.4  350.45  12.64  
 202     14.

In [9]:
x_train.shape,x_test.shape,y_train.shape,y_test.shape

((354, 13), (152, 13), (354,), (152,))

In [10]:
# selecting model
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [11]:
# train or fit model
model.fit(x_train,y_train)

LinearRegression()

In [12]:
model.intercept_

34.21916368862993

In [13]:
model.coef_

array([-1.29412069e-01,  3.65184937e-02,  1.54418944e-02,  2.35486887e+00,
       -2.04171489e+01,  4.41356565e+00,  4.61075512e-03, -1.58626723e+00,
        2.51478665e-01, -9.59591213e-03, -9.64169204e-01,  1.00972679e-02,
       -5.43198745e-01])

In [14]:
# predict model
y_pred = model.predict(x_test)

In [15]:
y_pred

array([31.71733828, 22.02143302, 21.16613197, 39.77837246, 20.10258512,
       22.86056216, 18.35574643, 14.7902735 , 22.55778646, 21.34594953,
       18.38491085, 27.9664665 , 29.85929012,  6.44680773, 10.68297311,
       26.24809521, 21.89368671, 25.22692365,  3.62385942, 36.21920372,
       24.07812335, 22.94103934, 14.27095261, 20.79013279, 24.22725035,
       16.7379611 , 18.74856986, 20.96709658, 28.513571  , 20.86346628,
        9.23450577, 17.06754852, 22.06953886, 22.23121875, 39.25875323,
       26.16769924, 42.50354003, 19.34517962, 34.51869058, 14.07023676,
       13.81055358, 23.27727535, 11.79100403,  9.01040731, 21.64587594,
       25.55339317, 18.16941728, 16.81991401, 14.66170215, 14.86477172,
       33.78924259, 33.26959074, 15.49208778, 24.08269034, 27.63531226,
       19.58288727, 45.02488529, 20.96959671, 20.07202649, 27.67146866,
       34.59154418, 12.71353064, 23.66247812, 31.65792337, 28.97459925,
       32.45963484, 13.93494747, 35.491924  , 19.35871482, 19.60

In [16]:
y_test


229    31.5
45     19.3
119    19.3
232    41.7
461    17.7
       ... 
47     16.6
39     30.8
132    23.0
169    22.3
278    29.1
Name: MEDV, Length: 152, dtype: float64

In [18]:
from sklearn.metrics import mean_absolute_error,mean_squared_error,mean_absolute_percentage_error
mean_absolute_percentage_error(y_test,y_pred)

0.1635593588221789

So,Our multiple regression model is of 16% error.that is its accuracy is 84%
